File: Assets/Script/Lib/UIX/context.ts
import { React } from "DoraX";
import { doraPrismTheme, mergeTheme, PartialTheme, Theme } from "UIX/theme";
import type { UiInputMode } from "UIX/types";
import { FocusManager } from "UIX/input/FocusManager";

export interface UiContext {
	theme: Theme;
	inputMode: UiInputMode;
	focusManager: FocusManager;
	scale: number;
}

const defaultFocusManager = new FocusManager();

let currentContext: UiContext = {
	theme: doraPrismTheme,
	inputMode: "pointer",
	focusManager: defaultFocusManager,
	scale: 1,
};

export function getUiContext(this: void): UiContext {
	return currentContext;
}

export interface UiProviderProps {
	theme?: PartialTheme;
	inputMode?: UiInputMode;
	scale?: number;
	children?: unknown;
}

export function UiProvider(this: void, props: UiProviderProps): React.Element | React.Element[] {
	currentContext = {
		theme: mergeTheme(doraPrismTheme, props.theme),
		inputMode: props.inputMode ?? "pointer",
		focusManager: currentContext.focusManager,
		scale: props.scale ?? 1,
	};
	return props.children as React.Element | React.Element[];
}

export interface ThemeScopeProps {
	theme: PartialTheme;
	children?: unknown;
}

export function ThemeScope(this: void, props: ThemeScopeProps): React.Element | React.Element[] {
	const previous = currentContext;
	currentContext = {
		theme: mergeTheme(previous.theme, props.theme),
		inputMode: previous.inputMode,
		focusManager: previous.focusManager,
		scale: previous.scale,
	};
	return props.children as React.Element | React.Element[];
}

File: Assets/Script/Lib/UIX/context.lua
-- [ts]: context.ts
local ____lualib = require("lualib_bundle") -- 1
local __TS__New = ____lualib.__TS__New -- 1
local ____exports = {} -- 1
local ____theme = require("UIX.theme") -- 2
local doraPrismTheme = ____theme.doraPrismTheme -- 2
local mergeTheme = ____theme.mergeTheme -- 2
local ____FocusManager = require("UIX.input.FocusManager") -- 4
local FocusManager = ____FocusManager.FocusManager -- 4
local defaultFocusManager = __TS__New(FocusManager) -- 13
local currentContext = {theme = doraPrismTheme, inputMode = "pointer", focusManager = defaultFocusManager, scale = 1} -- 15
function ____exports.getUiContext() -- 22
	return currentContext -- 23
end -- 22
function ____exports.UiProvider(props) -- 33
	currentContext = { -- 34
		theme = mergeTheme(doraPrismTheme, props.theme), -- 35
		inputMode = props.inputMode or "pointer", -- 36
		focusManager = currentContext.focusManager, -- 37
		scale = props.scale or 1 -- 38
	} -- 38
	return props.children -- 40
end -- 33
function ____exports.ThemeScope(props) -- 48
	local previous = currentContext -- 49
	currentContext = { -- 50
		theme = mergeTheme(previous.theme, props.theme), -- 51
		inputMode = previous.inputMode, -- 52
		focusManager = previous.focusManager, -- 53
		scale = previous.scale -- 54
	} -- 54
	return props.children -- 56
end -- 48
return ____exports -- 48